### PR TITLE
feat: add support for https only procotol in NEG backend service

### PR DESCRIPTION
**Pull request by gangsta**
As per discussion in https://github.com/terraform-google-modules/terraform-google-lb-http/issues/201

I have created basic changes to the module, please let me know if there is a better way of doing it.
I checked some different scenarios and come up with the following changes, which will allow
* to support only HTTPS or HTTP2 backend services.

I'm up for changes If this doesn't fit your current workstyle/module style.

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate . 
Success!
 The configuration is valid.
terraform_validate ./examples/cloudrun 
Success!
 The configuration is valid.
terraform_validate ./examples/https-gke 
Success!
 The configuration is valid.
terraform_validate ./examples/https-gke/gke-node-port 
Success!
 The configuration is valid.
terraform_validate ./examples/https-redirect 
Success!
 The configuration is valid.
terraform_validate ./examples/mig-nat-http-lb 
Success!
 The configuration is valid.
terraform_validate ./examples/multi-backend-multi-mig-bucket-https-lb 
Success!
 The configuration is valid.
terraform_validate ./examples/multi-mig-http-lb 
Success!
 The configuration is valid.
terraform_validate ./examples/multiple-certs 
Success!
 The configuration is valid.
terraform_validate ./examples/shared-vpc 
Success!
 The configuration is valid.
terraform_validate ./modules/dynamic_backends 
Success!
 The configuration is valid.
terraform_validate ./modules/serverless_negs 
There are some problems with the configuration, described below.
The Terraform configuration must be valid before initialization so that
Terraform can determine which modules and providers need to be installed.
Error: 
Invalid combination of "count" and "for_each"
  on main.tf line 159, in resource "google_compute_backend_service" "default":
 159:   
for_each
 = var.backends
The "count" and "for_each" meta-arguments are mutually-exclusive, only one
should be used to be explicit about the number of resources to be created.
Error: 
Invalid combination of "count" and "for_each"
  on main.tf line 203, in resource "google_compute_backend_service" "custom":
 203:   
for_each
 = var.backends
The "count" and "for_each" meta-arguments are mutually-exclusive, only one
should be used to be explicit about the number of resources to be created.
terraform_validate ./test/fixtures/cloudrun 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/https-redirect 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/mig_nat 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/multi_certs 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/multi_mig 
Success!
 The configuration is valid.
terraform_validate ./test/setup 
Success!
 The configuration is valid.
 ```

File: autogen/main.tf.tmpl [hunk 1]
```diff
@@ -165,10 +165,9 @@ resource "google_compute_backend_service" "default" {
   project = var.project
   name    = "${var.name}-backend-${each.key}"
 
-  {% if not serverless %}{# not necessary for serverless as default port_name=http, protocol=HTTP #}
-  port_name                       = each.value.port_name
-  protocol                        = each.value.protocol
-  {% endif %}
+  port_name = lookup(each.value, "port_name", "http")
+  protocol  = lookup(each.value, "protocol", "HTTP")
+
```
Comment (gangsta) [inline autogen/main.tf.tmpl:170]: I would love to have here extra :eyes: .
should we move this default to variables? or leave it here?

one more thing: each.value.port_name vs lookup(each.value, "port_name", "HTTP")

which one is better ?

**Review by morgante — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: autogen/variables.tf.tmpl [hunk 2]
```diff
@@ -84,11 +84,12 @@ variable "target_service_accounts" {
 variable "backends" {
   description = "Map backend indices to list of backend maps."
   type = map(object({
```
[the remark this answers is not in the document]
Comment (gangsta) [reply autogen/variables.tf.tmpl:86]: awesome let me fix it :+1: 

Thanks for helping out here

[the remark this answers is not in the document]
Comment (gangsta) [reply autogen/variables.tf.tmpl]: oh now I see what u refer to.
according to this: https://www.terraform.io/docs/language/expressions/types.html

block of 85-148 line should be changed to type map

is that correct ?

[the remark this answers is not in the document]
Comment (morgante) [reply autogen/variables.tf.tmpl]: It should be changed to `map(any)`.

[the remark this answers is not in the document]
Comment (gangsta) [reply autogen/variables.tf.tmpl]: I have added map(any) but ci is failing, is there a chance to see what is error so I can fix it?

[the remark this answers is not in the document]
Comment (morgante) [reply autogen/variables.tf.tmpl]: The error is commented above ^^^

**Comment by gangsta** [comment]
Hi @morgante .
here is example of setup we want to accomplish.

The current setup is by default creating `HTTP only` or `HTTP and HTTPS` protocols for NEG. But in my case, we need only `HTTPS only`.

config example is going to be 

```
module "neg_external" {
  source               = "GoogleCloudPlatform/lb-http/google//modules/serverless_negs"
  version              = "~> 5.1"
  name                 = "neg-external"
  project              = var.project_id
  ssl                  = true
  ssl_certificates     = ["projects/someproject-here/global/sslCertificates/wildcard-certificate"]
  use_ssl_certificates = true
  http_forward         = false
  create_address       = false
  address              = "someiphere"

  backends = {
    default = {
      description                     = null
      protocol                        = "HTTPS"
      port                            = 443
      port_name                       = "https"
      timeout_sec                     = 300
      connection_draining_timeout_sec = 60
      enable_cdn                      = false
      security_policy                 = google_compute_security_policy.allow_some_policy.self_link
      session_affinity                = null
      affinity_cookie_ttl_sec         = null
      custom_request_headers          = null
      custom_response_headers         = null

      log_config = {
        enable      = true
        sample_rate = null
      }

      groups = [
        {
          group                        = google_compute_region_network_endpoint_group.some_neg.id
          balancing_mode               = null
          capacity_scaler              = null
          description                  = null
          max_connections              = null
          max_connections_per_instance = null
          max_connections_per_endpoint = null
          max_rate                     = null
          max_rate_per_instance        = null
          max_rate_per_endpoint        = null
          max_utilization              = null
        }
      ]
      iap_config = {
        enable               = false
        oauth2_client_id     = ""
        oauth2_client_secret = ""
      }
    }
  }
}
```

**Comment by gangsta** [comment]
Still active

**Comment by kkirpichnikov** [comment]
@morgante any plans to review and merge it?

**Pull request closed without merging**